src/Tochka.JsonRpc.Client/Models/SingleJsonRpcResult.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Text.Json;
+using JetBrains.Annotations;
+using Tochka.JsonRpc.Common;
+using Tochka.JsonRpc.Common.Models.Response;
+using Tochka.JsonRpc.Common.Models.Response.Errors;
+using Tochka.JsonRpc.Common.Models.Response.Untyped;
+
+namespace Tochka.JsonRpc.Client.Models;
+
+[PublicAPI]
+public class SingleJsonRpcResult : ISingleJsonRpcResult
+{
+    private readonly IJsonRpcCallContext context;
+    private readonly JsonSerializerOptions headersJsonSerializerOptions;
+    private readonly JsonSerializerOptions dataJsonSerializerOptions;
+    private readonly IResponse? response;
+
+    public SingleJsonRpcResult(IJsonRpcCallContext context, JsonSerializerOptions headersJsonSerializerOptions, JsonSerializerOptions dataJsonSerializerOptions)
+    {
+        this.context = context ?? throw new ArgumentNullException(nameof(context));
+        if (context.BatchResponse != null)
+        {
+            throw new ArgumentOutOfRangeException(nameof(context), "Expected single response");
+        }
+
+        response = context.SingleResponse;
+        this.headersJsonSerializerOptions = headersJsonSerializerOptions;
+        this.dataJsonSerializerOptions = dataJsonSerializerOptions;
+    }
+
+    public T? GetResponseOrThrow<T>()
+    {
+        switch (response)
+        {
```

Comment:
switch expression instead?

Reply:
Wasn't possible because of `context.WithError` call.
Added chaining to context and changed it
✅